sdks/java/core/src/test/java/org/apache/beam/sdk/io/UnboundedReadFromBoundedSourceTest.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.RunnableOnService;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.Max;
+import org.apache.beam.sdk.transforms.Min;
+import org.apache.beam.sdk.transforms.RemoveDuplicates;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.PCollection;
+
+import com.google.api.client.util.Lists;
+import com.google.common.collect.Sets;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Unit tests for {@link UnboundedReadFromBoundedSource}.
+ */
+@RunWith(JUnit4.class)
+public class UnboundedReadFromBoundedSourceTest {
+
+  @Test
+  @Category(RunnableOnService.class)
+  public void testBoundedToUnboundedSourceAdapter() throws Exception {
+    long numElements = 100;
+    BoundedSource<Long> boundedSource = CountingSource.upTo(numElements);
+    UnboundedSource<Long, Checkpoint<Long>> unboundedSource =
+        new BoundedToUnboundedSourceAdapter<>(boundedSource);
+
+    Pipeline p = TestPipeline.create();
+
+    PCollection<Long> output =
+        p.apply(Read.from(unboundedSource).withMaxNumRecords(numElements));
+
+    // Count == numElements
+    PAssert
+      .thatSingleton(output.apply("Count", Count.<Long>globally()))
+      .isEqualTo(numElements);
+    // Unique count == numElements
+    PAssert
+      .thatSingleton(output.apply(RemoveDuplicates.<Long>create())
+                          .apply("UniqueCount", Count.<Long>globally()))
+      .isEqualTo(numElements);
+    // Min == 0
+    PAssert
+      .thatSingleton(output.apply("Min", Min.<Long>globally()))
+      .isEqualTo(0L);
+    // Max == numElements-1
+    PAssert
+      .thatSingleton(output.apply("Max", Max.<Long>globally()))
+      .isEqualTo(numElements - 1);
+    p.run();
+  }
+
+  @Test
+  public void testBoundedToUnboundedSourceAdapterCheckpoint() throws Exception {
+    long numElements = 100;
+    BoundedSource<Long> boundedSource = CountingSource.upTo(numElements);
+    BoundedToUnboundedSourceAdapter<Long> unboundedSource =
+        new BoundedToUnboundedSourceAdapter<>(boundedSource);
+
+    PipelineOptions options = PipelineOptionsFactory.create();
+    BoundedToUnboundedSourceAdapter<Long>.Reader reader =
+        unboundedSource.createReader(options, null);
+
+    Set<Long> expected = Sets.newHashSet();
+    for (long i = 0; i < numElements; ++i) {
+      expected.add(i);
+    }
+
+    List<Long> actual = Lists.newArrayList();
+    for (boolean hasNext = reader.start(); hasNext;) {
+      actual.add(reader.getCurrent());
+      if (actual.size() % 9 == 0) {
+        Coder<Checkpoint<Long>> checkpointCoder = unboundedSource.getCheckpointMarkCoder();
```

Comment:
add comments to explain what the test is doing. e.g., // checkpoint every 10 elements